File: 1_week/mario_hard.go
package main

import "fmt"

func main() {
  var num int
  fmt.Print("Enter a number: ")
  fmt.Scan(&num)
  for num<=0{
    fmt.Print("Invalid number: ")
    fmt.Scan(&num)
  }

  build_pyramid(num);
}

func build_pyramid(num int){
  for i := 0; i <num; i++ {
    for j := 0; j < num; j++ {
      build_left(num, i, j);
    }

    build_space();

    for j := 0; j < num; j++{
      build_right(i, j);
    }
    fmt.Printf("\n");
  }
}


func build_left(num, i, j int){
  if j < num-i-1{
    fmt.Printf(" ");
  }else {
    fmt.Printf("#");
  }
}

func build_right(i, j int){
  if(j < i+1){
    fmt.Printf("#");
  }else {
    fmt.Printf(" ");
  }
}

func build_space(){
  fmt.Printf("  ");
}

File: 1_week/mario_hard.c
#include <stdio.h>

void build_left(int num,int i, int j);
void build_right(int i, int j);
void build_space(void);
void build_pyramid(int num);

void main(void) {
  int num;
  printf("Enter a number: ");
  scanf("%i", &num);
  while (num<=0){
    printf("Invalid number: ");
    scanf("%i", &num);
  }

  build_pyramid(num);
}

void build_pyramid(int num){
  for(int i = 0; i <num; i++){
    for(int j = 0; j < num; j++){
      build_left(num, i, j);
    }

    build_space();

    for(int j= 0; j < num; j++){
      build_right(i, j);
    }
    printf("\n");
  }
}

void build_left(int num, int i, int j){
  if(j < num-i-1){
    printf(" ");
  }else {
    printf("#");
  }
}

void build_right(int i, int j){
  if(j < i+1){
    printf("#");
  }else {
    printf(" ");
  }
}

void build_space(void){
  printf("  ");
}

